File: 60. Permutation Sequence/optimized.py
class Solution:
    def getPermutation(self, n: int, k: int) -> str:
        fact = 1
        num = []
        for i in range(1,n):
            num.append(i)
            fact = fact*i
        num.append(n)
        k -= 1
        ans = ""
        while True:
            ans += str(num[k//fact])
            num.pop(k//fact)
            if len(num) == 0:
                break
            k = k % fact
            fact = fact // len(num)
        return ans

File: 60. Permutation Sequence/optimized.cpp
class Solution {
public:
    string getPermutation(int n, int k) {
        int fact = 1;
        vector<int> num;
        for(int i=1; i<n; i++){
            num.push_back(i);
            fact = fact*i;
        }
        num.push_back(n);
        string ans = "";
        k = k-1;
        while(true){
            ans += to_string(num[k/fact]);
            num.erase(num.begin()+(k/fact));
            if(num.size()==0){
                break;
            }
            k = k % fact;
            fact = fact / num.size();
        }
        return ans;
    }
};
